### DISCUSSION TITLE
Popper inconsistently will flip at bottom of scrolled table

### DISCUSSION THREAD

**Discussion by okbytes** [floating-ui/floating-ui#1397] [Q&A]
I have a scrollable table with a max-height of 400px. I have some select elements that allow the user to change a value. Popper is occasionally not flipping and will push out extra space for itself at the bottom of the table. It looks like this:

![image](https://user-images.githubusercontent.com/6867344/139743101-efb067cf-c88c-4602-b88b-411b3db7f550.png)

However, _sometimes_ Popper will do what I expect and flip. It looks like this:

![image](https://user-images.githubusercontent.com/6867344/139743188-004030c3-b25e-423d-8c99-3e3c17372cea.png)

My options object to `usePopper` looks like this:

```js
modifiers: [
            {name: "offset", options: {offset: [0, 8]}},
            {
                name: "preventOverflow",
                options: {
                    mainAxis: true, // true
                    altAxis: false, // false
                    padding: 0, // 0
                    boundary: "clippingParents", // "clippingParents"
                    altBoundary: false, // false
                    rootBoundary: "viewport", // "viewport"
                    tether: false, // true
                    tetherOffset: 0, // 0
                },
            },
        ],
```

I have no idea what's going on here or how to debug it.

**Comment by ulken** [comment]
I'm experiencing something similar where the popper element overflows the viewport on initial pop, but as soon as I scroll down, it correctly flips its position. _Even_ if I'm already scrolled to the bottom. 

Succeeding pops are positioned correctly. As long as I don't scroll again. At certain scroll levels it incorrectly overflows the viewport again. Really weird behavior.

If not clear from the original description, this is within a scroll container.

**Suboptimal workaround**

Right now, I'm explicitly calling `update()` after rendering to have it positioned correctly. However, it causes flickering since it's wrongly positioned at first.

**Reply by ulken** [reply]
I should also mention I use `react-popper`, but if I'm not totally missing something, all positioning logic should be in the core lib.

**Reply by atomiks** [reply]
Regarding the workaround, if you call `forceUpdate`, does it flicker? It should be called before the browser paints

As for the issue here, we'd need a reproduction to debug what's happening

**Reply by okbytes** [reply]
I had tried using `forceUpdate`, but to no success. I am trying to get a reproduction working on CodeSandbox. It's been difficult as even in my codebase, the problem is not consistently reproducible. I'll keep trying. In the mean time, I used a different library (`react-select`) and enabled locked scrolling on the body when the dropdown menu is open.

**Reply by atomiks** [reply]
@01brett are you appending the popper to the table or outside in the `<body>` / Portal? If you try with the core lib rather than `react-popper`, does the issue still occur? 

I feel like it has something to do with the popper getting mounted to the body, pushing out space, then being positioned, but since it pushed out the space, it doesn't need to flip anymore. But the initial `position: absolute; top: 0; left: 0;` styles should not make that occur...

**Reply by okbytes** [reply]
@atomiks I am adding `usePopper` to the listbox (headlessui/react) component's menu, which is inside the table. I will try your suggestion of using purely the core lib. When I tried using reachui's listbox, they used a portal on the body, which avoided this problem. However, there were other issues with that component (stale value state, no way to close menu when button is not in view) that led me to reach for `react-select` and unblock my co-workers.

**Reply by ulken** [reply]
I am also using the hook with the listbox component from Headless UI. I am not using a portal either as I want the width to be the same as the trigger. I know it can be calculated with JS, but I'd rather not.

Can enter transitions affect this maybe?

**Reply by ulken** [reply]
> Regarding the workaround, if you call forceUpdate, does it flicker? It should be called before the browser paints

I got rid of the flickering. That was my fault. I unnecessarily called it in a `useEffect()` which triggers on next render, so no wonder it flickered.
